### PR TITLE
Remember test answers

### PR DESCRIPTION
**Pull request by Fistulator**
Solving issue #360

### CODE REVIEW

File: web/src/main/frontend/src/components/TimeRemaining.js [hunk 1]
```diff
@@ -5,6 +5,7 @@ import {inject, observer} from 'mobx-react';
 @observer
 class TimeRemaining extends React.Component {
   componentWillUpdate() {
+    // TODO write onwarning method
```
Comment (stanislavgrujic) [inline web/src/main/frontend/src/components/TimeRemaining.js]: Do not leave hanging TODOs in code. Either create a task for it and reference it in the code, or clean it up right away.

File: web/src/main/java/eu/execom/testgenerator/web/dto/TestDto.java [hunk 2]
```diff
@@ -33,10 +33,14 @@ public TestDto(Test test) {
...
       this.questions = test.getTestQuestions()
                            .stream()
-                           .map(TestQuestion::getQuestion)
-                           .map(QuestionDto::new)
+                           .map(this::makeQuestionWithTestAnswers)
                            .collect(Collectors.toList());
     }
   }
 
+  private QuestionDto makeQuestionWithTestAnswers(TestQuestion testQuestion) {
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/TestDto.java]: More expressive would be to name the method `toQuestionWithTestAnswers` or `mapToQuestionWithTestAnswers`.

**Comment by stanislavgrujic** [comment]
@iscekic Please review frontend.

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js [hunk 3]
```diff
@@ -10,7 +10,21 @@ import {inject, observer} from 'mobx-react';
 class TakeTestQuestionComponent extends React.Component {
   //TODO: review iscekic, These two lines look like they belong in a lifecycle method.
   answers = this.props.newTestForm.$('answers').add();
+
+  prepareMultipleChoiceQuestion() {
+    const checkedAnswers = [...this.props.testAnswers.testQuestionAnswerIds];
```
Comment (iscekic) [inline web/src/main/frontend/src/components/TakeTestQuestionComponent.js]: There is no need to create a copy as we aren't mutating anything.

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js [hunk 4]
```diff
@@ -10,7 +10,21 @@ import {inject, observer} from 'mobx-react';
...
+    const allAnswerIds = [...this.props.question.answers].map(obj => obj.id);
```
Comment (iscekic) [inline web/src/main/frontend/src/components/TakeTestQuestionComponent.js]:
```suggestion
    const allAnswerIds = this.props.question.answers.map(obj => obj.id);
```

`map` returns a new array without touching the mapped one.

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js [hunk 5]
```diff
@@ -10,7 +10,21 @@ import {inject, observer} from 'mobx-react';
...
+    const andswersAndStatus = allAnswerIds.reduce((accumulator, id) => {
+      const isInCheckedAnswers = checkedAnswers.indexOf(id) !== -1;
+      accumulator[id] = isInCheckedAnswers;
```
Comment (iscekic) [inline web/src/main/frontend/src/components/TakeTestQuestionComponent.js]: You can inline this variable.

File: web/src/main/frontend/src/components/ui/Checkbox.js [hunk 6]
```diff
@@ -15,6 +15,8 @@ const Checkbox = props => {
     }
   }
 
+  propsContainer.defaultChecked = props.defaultChecked;
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ui/Checkbox.js]: I dislike this entire component. Can you refactor it to be more in line with `mobx-react-form`:

```js
// example usage
<Checkbox {...form.$('whatever').bind()} someProp="1" />
```

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js [hunk 7]
```diff
@@ -10,7 +10,21 @@ import {inject, observer} from 'mobx-react';
 class TakeTestQuestionComponent extends React.Component {
   //TODO: review iscekic, These two lines look like they belong in a lifecycle method.
```
Comment (iscekic) [inline web/src/main/frontend/src/components/TakeTestQuestionComponent.js]: Take a look at this `TODO` and see if you can fix it as part of this task.

File: web/src/main/frontend/src/components/TakeMultipleChoiceQuestion.js [hunk 8]
```diff
@@ -7,18 +7,14 @@ import Checkbox from './ui/Checkbox';
...
-      [option.id]: this.props.isSelected
-    });
     return (
       <Checkbox
         key={option.id}
         label={option.description}
         formField={answer.bind({
           value: option.id
         })}
+        defaultChecked={answer.values()[option.id]}
```
Comment (iscekic) [inline web/src/main/frontend/src/components/TakeMultipleChoiceQuestion.js]: See comment above. 💃

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js (hunk 3 above)
Comment (Fistulator) [reply web/src/main/frontend/src/components/TakeTestQuestionComponent.js]: I needed an array of checked answers, using .values() on testQuestionAnswerIds gives me an iterator, using it without values() gives me a much larger object than needed.

File: web/src/main/frontend/src/components/ui/Checkbox.js (hunk 6 above)
Comment (Fistulator) [reply web/src/main/frontend/src/components/ui/Checkbox.js]:
```suggestion
  <Checkbox
        answer = {answer}
        key = {option.id}
        option={option}
      />
```
Would something like this be better? Checkbox's code would look like:

```suggestion
<div className="checkbox_wrapper">
      <label>
        <input {...props.answer.bind(
          {
            value: props.option.id
            })} 
            defaultChecked={checked} 
            type="checkbox" />
        <div className="label_icon">
          <span className={answerTextClassName}>{props.option.description}</span>
        </div>
      </label>
    </div>
```

Comment (Fistulator) [reply web/src/main/frontend/src/components/ui/Checkbox.js]: Not sure I understand what I need to do.

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js (hunk 3 above)
Comment (iscekic) [reply web/src/main/frontend/src/components/TakeTestQuestionComponent.js]: My comment was about not copying `testQuestionAnswerIds` into a new variable `checkedAnswers` as we aren't changing it anywhere. No change == no need to copy.

File: web/src/main/frontend/src/components/ui/Checkbox.js (hunk 6 above)
Comment (iscekic) [reply web/src/main/frontend/src/components/ui/Checkbox.js]: I'd like the checkbox component to behave as much as eg. a simple input would:

```js
// https://foxhound87.github.io/mobx-react-form/docs/quick-start.html
<input {...form.$('whatever').bind()} />
```

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js (hunk 3 above)
Comment (Fistulator) [reply web/src/main/frontend/src/components/TakeTestQuestionComponent.js]: Understood.

File: web/src/main/frontend/src/components/ui/Checkbox.js (hunk 6 above)
Comment (Fistulator) [reply web/src/main/frontend/src/components/ui/Checkbox.js]: Upon looking further at the code, I realized that CheckBox was used by two different forms and that their functionality was relatively different. I decided to remove the component entirely as it was not doing much already and making it work with different form types would require either adding an addiotional component type that would be used only once or making some ugly code to make it comply with everything.

File: web/src/main/frontend/src/components/Test.js [hunk 9]
```diff
@@ -55,6 +55,8 @@ class Test extends React.Component {
   //TODO vrackovic: Investigate implementation and fill this function if needed
   handleAlarm = () => {};
 
+  handleWarning = () => {};
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Test.js]: Should this be empty without a `// TODO`?

Comment (Fistulator) [reply web/src/main/frontend/src/components/Test.js]: Will write an issue for these functions.

File: web/src/main/java/eu/execom/testgenerator/web/dto/TestQuestionAnswerDto.java [hunk 10]
```diff
@@ -16,7 +15,7 @@
   private String input;
   private List<Long> testQuestionAnswerIds;
 
-  TestQuestionAnswerDto(TestQuestion testQuestion) {
+  public TestQuestionAnswerDto(TestQuestion testQuestion) {
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/TestQuestionAnswerDto.java]: Why did you change the access modifier? It is not used outside the dto package, as far as I see it?

Comment (Fistulator) [reply web/src/main/java/eu/execom/testgenerator/web/dto/TestQuestionAnswerDto.java]: I believe it caused an error, but the access modifier could not have been the cause (as everything works as it should after removing the access modifier)

Restored it to its original value now.

**Comment by Fistulator** [comment]
@iscekic has already reviewed the frontend part.
